### PR TITLE
Feat/dynamite/reduce-name-collisions

### PR DESCRIPTION
**Pull request by Leptopoda**
Signed-off-by: Nikolas Rimikis <leptopoda@users.noreply.github.com>
fixes: #1154

### CODE REVIEW

File: packages/dynamite/dynamite/lib/src/builder/serializer.dart [hunk 1]
```diff
@@ -5,15 +5,22 @@ List<Spec> buildSerializer(final State state) => [
...
           ..type = refer('Serializers')
-          ..name = 'serializers';
+          ..name = r'$serializers'
+          ..assignment = const Code(r'_$serializers');
+      }),
+      Field((final b) {
+        b
+          ..modifier = FieldModifier.final$
+          ..type = refer('Serializers')
+          ..name = r'_$serializers';
```
Comment (provokateurin) [inline packages/dynamite/dynamite/lib/src/builder/serializer.dart:23]: Why have a public and a private variable that are the same? One public one with the $ prefix is enough

File: packages/dynamite/dynamite/lib/src/builder/serializer.dart [hunk 2]
```diff
@@ -31,15 +38,22 @@ List<Spec> buildSerializer(final State state) => [
...
           ..type = refer('Serializers')
-          ..name = 'jsonSerializers';
+          ..name = r'$jsonSerializers'
+          ..assignment = const Code(r'_$jsonSerializers');
+      }),
+      Field((final b) {
+        b
+          ..modifier = FieldModifier.final$
+          ..type = refer('Serializers')
+          ..name = r'_$jsonSerializers';
```
Comment (provokateurin) [inline packages/dynamite/dynamite/lib/src/builder/serializer.dart:56]: Same as above

**Review by provokateurin — CHANGES_REQUESTED** [review]

File: packages/dynamite/dynamite/lib/src/builder/serializer.dart (hunk 1 above)
Comment (Leptopoda) [reply packages/dynamite/dynamite/lib/src/builder/serializer.dart:23]: Nope it isn't:
In the method body of an operation we need private variables.
Imagine an operation with a parameter called `jsonSerializers`:

```dart
  DynamiteRawResponse<JsonObject, void> $getRaw({
    String? jsonSerializers,
  }) {
    final _parameters = <String, dynamic>{};
    const _headers = <String, String>{
      'Accept': 'application/json',
    };

    final $jsonSerializers = jsonSerializers.serialize( // need a private serializer here
      contentString,
      specifiedType: const FullType(ContentString, [
        FullType(BuiltMap, [FullType(String), FullType(JsonObject)]),
      ]),
    );
    _parameters['json-serializers'] = $jsonSerializers;
    
    // ...
  }
```

Comment (provokateurin) [reply packages/dynamite/dynamite/lib/src/builder/serializer.dart:23]: But something like
```dart
    final _$jsonSerializers = $jsonSerializers.serialize( // need a private serializer here
      jsonSerializers,
      specifiedType: const FullType(String),
    );
```
should work.
The serialized name just needs to be prefixed with underscore.

Comment (Leptopoda) [reply packages/dynamite/dynamite/lib/src/builder/serializer.dart:23]: but then a parameter called `headers` or `path` (we already faced the later one once) would conflict with our already private variables.

Yes we can prefix them with `_$` but I don't know if we should rework such a major thing now.

Comment (provokateurin) [reply packages/dynamite/dynamite/lib/src/builder/serializer.dart:23]: Alright then

**Review by provokateurin — CHANGES_REQUESTED** [review]
Please make the public serializers getters pointing to the private serializers.

**Comment by Leptopoda** [comment]
I'll add an e2e test for both collision types we discussed above. Just in case we forget this in the future.
 
 > Please make the public serializers getters pointing to the private serializers.

What do you mean?

**Comment by provokateurin** [comment]
> What do you mean?

```dart
Serializers get $jsonSerializers => _$jsonSerializers;
final Serializers _$jsonSerializers = (_$serializers.toBuilder()
```

Now that I think about it, this isn't really fixing the problem and rather working around it. In case there is a parameter called `jsonSerializers` the `$jsonSerializers` will be declared twice. Of course it works, but it's rather ugly imo

**Comment by Leptopoda** [comment]
Yes we rely on the parameter shadowing the library level variable.
Any better ideas?

**Comment by provokateurin** [comment]
Other than what I said above, no. Go ahead then

**Comment by Leptopoda** [comment]
If you want to change it please open an issue. I'm not against prefixing our stuff with a `$` and the serialized variables with a `_`.

I'd just like to think more about the implications and not rush such a change with this PR.

**Review by provokateurin — APPROVED** [review]

**Pull request merged**